src/hotspot/share/cds/classListWriter.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "cds/classListWriter.hpp"
+#include "classfile/classFileStream.hpp"
+#include "classfile/classLoader.hpp"
+#include "classfile/classLoaderData.hpp"
+#include "classfile/moduleEntry.hpp"
+#include "classfile/systemDictionaryShared.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/instanceKlass.hpp"
+#include "runtime/mutexLocker.hpp"
+
+fileStream* ClassListWriter::_classlist_file = NULL;
+
+void ClassListWriter::init() {
+  // For -XX:DumpLoadedClassList=<file> option
+  if (DumpLoadedClassList != NULL) {
+    const char* list_name = make_log_name(DumpLoadedClassList, NULL);
+    _classlist_file = new(ResourceObj::C_HEAP, mtInternal)
+                         fileStream(list_name);
+    _classlist_file->print_cr("# NOTE: Do not modify this file.");
+    _classlist_file->print_cr("#");
+    _classlist_file->print_cr("# This file is generated via the -XX:DumpLoadedClassList=<class_list_file> option");
+    _classlist_file->print_cr("# and is used at CDS archive dump time (see -Xshare:dump).");
+    _classlist_file->print_cr("#");
+    FREE_C_HEAP_ARRAY(char, list_name);
+  }
+}
+
+void ClassListWriter::write(const InstanceKlass* k, const ClassFileStream* cfs) {
+  assert(is_enabled(), "must be");
```

Comment:
Should `assert_locked()` be added here?

Reply:
This function doesn't need the lock, since it doesn't directly use `ClassListWriter::_classlist_file` or `ClassListWriter::_id_table`, which are protected by `ClassListFile_lock`. The lock will be taken a few lines down, in the `ClassListWriter` constructor.

Your comment reminded me to look for places that need the lock. As a result, I added `assert_locked()` to `ClassListWriter::write_to_stream()`.